### DISCUSSION TITLE
How do I require an Any?

### DISCUSSION THREAD

**Discussion by ericzbeard** [apple/pkl#316] [Q&A] [answered]
I need to define a class with a required property that can be anything, but `Any` is apparently equivalent to `Any?`. This code fails, which is confusing because if I replace `hidden Foo: Any` with `hidden Foo: Any?`, it succeeds:

```pkl
open class Test {
    hidden Foo: Any

    // This works
    // hidden Foo: Any?

    hidden Str: String
    
    Props {
        ["Foo"] = Foo
        ["Str"] = Str
    }
}

t = new Test {
    Foo {
        Bar = "Baz"
    }

    // This works
    // Foo = ""

    Str = ""
}
```

**Comment by ericzbeard** [comment]
Ok I think I figured this out.. just need to replace `Any` with `Dynamic`.

**Comment by bioball** [comment]
You want `foo: NonNull`; `NonNull` is defined as:

```
typealias NonNull = Any(!(this is Null))
```

**Reply by ericzbeard** [reply]
What's the difference between that and Dynamic?

**Reply by bioball** [reply]
`Dynamic` is its own leaf type, and doesn't permit other types. On the other hand, `Any` is the top type; it permits any value. `NonNull` is basically just `Any` that doesn't permit `Null`.

The following throws:

```
foo: Dynamic = 1
```

But this will succeed:

```
foo: NonNull = 1
```

`Dynamic` is a special type of object that allows any property, element, or entry. It's actually rarely used in normal Pkl. It's more common to define your data in terms of classes, mappings, and listings, which will give you validation. `Dynamic` is useful as an ad-hoc data type if you want to just start writing Pkl without the overhead of first defining a schema.

More details: https://pkl-lang.org/main/current/language-reference/index.html#typed-objects

**Reply by ericzbeard** [reply, marked as answer]
I think Dynamic is what I need. I'm modeling AWS CloudFormation registry schemas, which sometimes define properties as "object", which can be any valid YAML/JSON.

**Reply by ericzbeard** [reply]
The sample code still fails with NonNull

**Reply by bioball** [reply]
Gotcha!

For CloudFormation: we haven't published it anywhere, but: I've played around with using our [JSON Schema generator](https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/org.json_schema.contrib/current/generate/index.html) to generate Pkl modules, and it works pretty well.

Here's a gist. You can clone this down somewhere, run `bash download-schemas.sh`, then run `pkl eval convert.pkl` to generate the Pkl output.

https://gist.github.com/bioball/9edf23a41695cb2080131ec3610a4c17

**Reply by ericzbeard** [reply]
Nice! I'll check that out. I already have schema processing for CloudFormation in [Rain](https://github.com/aws-cloudformation/rain), working on pkl support in [this PR](https://github.com/aws-cloudformation/rain/pull/318). Fun project!

**Reply by bioball** [reply]
Awesome! That looks really cool!
